core/src/test/java/org/apache/brooklyn/feed/http/HttpFeedTest.java

```diff
@@ -360,7 +364,90 @@ public void testPollsMultiClearsOnSubsequentFailure() throws Exception {
         
         server.shutdown();
     }
+    
+    @Test
+    public void testPreemptiveBasicAuth() throws Exception {
+        final String username = "brooklyn";
+        final String password = "hunter2";
+
+        feed = HttpFeed.builder()
+                .entity(entity)
+                .baseUrl(server.getUrl("/"))
+                .credentials(username, password)
+                .preemptiveBasicAuth(true)
+                .poll(new HttpPollConfig<Integer>(SENSOR_INT)
+                        .period(100)
+                        .onSuccess(HttpValueFunctions.responseCode())
+                        .onException(Functions.constant(-1)))
+                .build();
+
+        EntityAsserts.assertAttributeEqualsEventually(entity, SENSOR_INT, 200);
+        RecordedRequest req = server.takeRequest();
+        String headerVal = req.getHeader("Authorization");
+        String expectedVal = getBasicAuthHeaderVal(username, password);
+        assertEquals(headerVal, expectedVal);
+    }
+
+    @Test
+    public void testPreemptiveBasicAuthFailsIfNoCredentials() throws Exception {
+        try {
+            feed = HttpFeed.builder()
+                    .entity(entity)
+                    .baseUrl(new URL("http://shouldNeverBeCalled.org"))
+                    .preemptiveBasicAuth(true)
+                    .poll(new HttpPollConfig<Integer>(SENSOR_INT)
+                            .period(100)
+                            .onSuccess(HttpValueFunctions.responseCode())
+                            .onException(Functions.constant(-1)))
+                    .build();
+            Asserts.shouldHaveFailedPreviously();
+        } catch (IllegalArgumentException e) {
+            Asserts.expectedFailureContains(e, "Must not enable preemptiveBasicAuth when there are no credentials");
+        }
+    }
 
+    // Expected behaviour of o.a.http.client is that it first sends the request without credentials,
+    // and then when given a challenge for basic-auth it re-sends the request with the basic-auth header.
+    @Test
+    public void testNonPreemptiveBasicAuth() throws Exception {
+        final String username = "brooklyn";
+        final String password = "hunter2";
+        
+        if (server != null) server.shutdown();
```

Comment:
The mock server stuff in this file is in a bit of a mess.  I'd suggest we clean it up but not in this PR.